src/library/scala/collection/immutable/Vector.scala

```diff
@@ -402,14 +441,14 @@ final class Vector[+A] private[immutable] (private[collection] val startIndex: I
     }
   }
 
-  private def copyLeft(array: Array[AnyRef], right: Int): Array[AnyRef] = {
-    val copy = new Array[AnyRef](array.length)
-    java.lang.System.arraycopy(array, 0, copy, 0, right)
+  private def copyLeft[T <: AnyRef](array: Array[T], right: Int): Array[T] = {
+    val copy = array.clone()
+    java.util.Arrays.fill(copy.asInstanceOf[Array[AnyRef]], right, array.length, null)
```

Comment:
Wouldn't `Arrays.copyOf` be more efficient?

Reply:
I think that copyOf and clone are about the same code. I will test unless @retronym knows

Reply:
`Array.clone` or `Arrays.copyOf` should have near identical performance here AFAIK. `copyOf` is more general because you can specify a different length and different component type. Both avoid zeroing of the part of the new array that will be filled with copied elements.

`new Array; System.arrayCopy` can also avoid zeroing, so long as the bytecode has a shape that JIT can recognize.

So the following _might_ be better still:

```scala
    val copy = if (array.getClass.getComponentType eq classOf[AnyRef]) new Array[AnyRef](array.length) else Arrays.newInstance(array.length, array.getClass.getComponentType)
    java.lang.System.arraycopy(array, 0, copy, 0, right)
```

A microbenchmark of the options and analysis of JITted code would help.

Reply:
Oh, the element type is always `AnyRef`. @mkeskells What was the motivation to use `clone`? The commit message doesn't explain that.

Reply:
Its a problem with the java/scala interpretation of variance as I see it
`val copy = util.Arrays.copyOf(array)` doesn't compile
```
Error:(446, 28) overloaded method value copyOf with alternatives:
  (x$1: Array[Boolean],x$2: Int)Array[Boolean] <and>
  (x$1: Array[Double],x$2: Int)Array[Double] <and>
  (x$1: Array[Float],x$2: Int)Array[Float] <and>
  (x$1: Array[Char],x$2: Int)Array[Char] <and>
  (x$1: Array[Long],x$2: Int)Array[Long] <and>
  (x$1: Array[Int],x$2: Int)Array[Int] <and>
  (x$1: Array[Short],x$2: Int)Array[Short] <and>
  (x$1: Array[Byte],x$2: Int)Array[Byte] <and>
  [T, U](x$1: Array[U with Object], x$2: Int, x$3: Class[_ <: Array[T with Object]])Array[T with Object] <and>
  [T](x$1: Array[T with Object], x$2: Int)Array[T with Object]
 cannot be applied to (Array[T])
    val copy = util.Arrays.copyOf(array)
```

Reply:
Okay. I think your solution with `clone()` is probably okay because these arrays are tiny, so we might not see the effect of copying the trailing elements before nulling them out.

But the following is probably the smallest delta from the original code, all it changes is making your intended change to strongly type the arrays to avoid the need for a `checkcast` on the elements after reading out of `display{1-5}`.

```scala
  private def copyLeft[T <: AnyRef](array: Array[T], right: Int): Array[T] = {
    val copy = 
      if (array.getClass.getComponentType eq classOf[AnyRef]) new Array[AnyRef](array.length) 
      else java.lang.reflect.Array.newInstance(array.getClass.getComponentType, array.length)
    java.lang.System.arraycopy(array, 0, copy, 0, right)
    copy.asInstanceOf[Array[T]]
  }
```

Reply:
happy to change to this if you prefer
I guess it depends if more elements are  copied or nulled, is `right` > half the array length. 
I don't believe that the bytecode will elide the nulling that is not needed, so we either copy all the data then remove some or start with null filled array and copy the data on top

why would we switch on getComponentType, rather than just 
`val copy = java.lang.reflect.Array.newInstance(array.getClass.getComponentType, array.length)`

I have not benchmarked what is faster though

Reply:
clone seems fastest based on the tst below, with some reasoning on the high error bars
```
[info] Benchmark                        (anyRef)  (right)  Mode  Cnt   Score    Error  Units
[info] ArrayCopyingBenchmark.copyLeft1      true        0  avgt   20  27.039 ▒  1.366  ns/op
[info] ArrayCopyingBenchmark.copyLeft1      true        8  avgt   20  33.348 ▒  1.898  ns/op
[info] ArrayCopyingBenchmark.copyLeft1      true       16  avgt   20  32.387 ▒  1.185  ns/op
[info] ArrayCopyingBenchmark.copyLeft1      true       24  avgt   20  32.245 ▒  1.224  ns/op
[info] ArrayCopyingBenchmark.copyLeft1      true       32  avgt   20  27.948 ▒ 16.927  ns/op
[info] ArrayCopyingBenchmark.copyLeft1     false        0  avgt   20  47.638 ▒ 40.166  ns/op
[info] ArrayCopyingBenchmark.copyLeft1     false        8  avgt   20  48.488 ▒ 28.803  ns/op
[info] ArrayCopyingBenchmark.copyLeft1     false       16  avgt   20  43.401 ▒ 30.312  ns/op
[info] ArrayCopyingBenchmark.copyLeft1     false       24  avgt   20  31.992 ▒  1.610  ns/op
[info] ArrayCopyingBenchmark.copyLeft1     false       32  avgt   20  35.233 ▒  2.392  ns/op
[info] ArrayCopyingBenchmark.copyLeft2      true        0  avgt   20  27.059 ▒  1.500  ns/op
[info] ArrayCopyingBenchmark.copyLeft2      true        8  avgt   20  31.948 ▒  1.812  ns/op
[info] ArrayCopyingBenchmark.copyLeft2      true       16  avgt   20  32.710 ▒  1.910  ns/op
[info] ArrayCopyingBenchmark.copyLeft2      true       24  avgt   20  32.589 ▒  1.839  ns/op
[info] ArrayCopyingBenchmark.copyLeft2      true       32  avgt   20  31.938 ▒  1.040  ns/op
[info] ArrayCopyingBenchmark.copyLeft2     false        0  avgt   20  25.823 ▒  0.700  ns/op
[info] ArrayCopyingBenchmark.copyLeft2     false        8  avgt   20  31.212 ▒  1.005  ns/op
[info] ArrayCopyingBenchmark.copyLeft2     false       16  avgt   20  31.737 ▒  1.104  ns/op
[info] ArrayCopyingBenchmark.copyLeft2     false       24  avgt   20  32.127 ▒  1.244  ns/op
[info] ArrayCopyingBenchmark.copyLeft2     false       32  avgt   20  32.172 ▒  1.259  ns/op
[info] ArrayCopyingBenchmark.copyLeft3      true        0  avgt   20  33.373 ▒  4.093  ns/op
[info] ArrayCopyingBenchmark.copyLeft3      true        8  avgt   20  51.340 ▒ 28.107  ns/op
[info] ArrayCopyingBenchmark.copyLeft3      true       16  avgt   20  24.208 ▒  2.127  ns/op
[info] ArrayCopyingBenchmark.copyLeft3      true       24  avgt   20  21.332 ▒  1.664  ns/op
[info] ArrayCopyingBenchmark.copyLeft3      true       32  avgt   20  16.394 ▒  1.315  ns/op
[info] ArrayCopyingBenchmark.copyLeft3     false        0  avgt   20  33.491 ▒  3.056  ns/op
[info] ArrayCopyingBenchmark.copyLeft3     false        8  avgt   20  27.359 ▒  1.845  ns/op
[info] ArrayCopyingBenchmark.copyLeft3     false       16  avgt   20  24.064 ▒  1.672  ns/op
[info] ArrayCopyingBenchmark.copyLeft3     false       24  avgt   20  22.240 ▒  2.261  ns/op
[info] ArrayCopyingBenchmark.copyLeft3     false       32  avgt   20  15.988 ▒  1.314  ns/op
```

```
package scala.collection

import java.util.concurrent.TimeUnit

import org.openjdk.jmh.annotations._

import scala.util.Random

@BenchmarkMode(Array(Mode.AverageTime))
@Fork(2)
@Threads(1)
@Warmup(iterations = 10)
@Measurement(iterations = 10)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
class ArrayCopyingBenchmark {
  val arraySize = 32

  @Param(Array("0", "8", "16", "24", "32"))
  var right: Int = _
  @Param(Array("true", "false"))
  var anyRef: Boolean = _

  var values: Array[_ <: AnyRef] = _


  @Setup(Level.Trial) def initKeys(): Unit = {
    values = if (anyRef) new Array[AnyRef](arraySize) else new Array[List[Int]](arraySize)
  }

  @Benchmark def copyLeft1: Array[_] = {
    copyLeft1(values, right)
  }
  private def copyLeft1[T <: AnyRef](array: Array[T], right: Int): Array[T] = {
    val copy =
      if (array.getClass.getComponentType eq classOf[AnyRef]) new Array[AnyRef](array.length)
      else java.lang.reflect.Array.newInstance(array.getClass.getComponentType, array.length)
    java.lang.System.arraycopy(array, 0, copy, 0, right)
    copy.asInstanceOf[Array[T]]
  }
  @Benchmark def copyLeft2: Array[_] = {
    copyLeft2(values, right)
  }
  private def copyLeft2[T <: AnyRef](array: Array[T], right: Int): Array[T] = {
    val copy = java.lang.reflect.Array.newInstance(array.getClass.getComponentType, array.length)
    java.lang.System.arraycopy(array, 0, copy, 0, right)
    copy.asInstanceOf[Array[T]]
  }
  @Benchmark def copyLeft3: Array[_] = {
    copyLeft3(values, right)
  }
  private def copyLeft3[T <: AnyRef](array: Array[T], right: Int): Array[T] = {
    val copy = array.clone()
    java.util.Arrays.fill(copy.asInstanceOf[Array[AnyRef]], right, array.length, null)
    copy
  }
}
```

Reply:
I could reproduce your results that show the clone is faster. For larger arrays it is quite a bit slower so this isn't an intuitive result!

Reply:
I was surprised by the result as well. I didn't dig though